severn/api/requirements.py

```diff
@@ -0,0 +1,282 @@
+        self.beta = b or MAX_VERSION
+        self.rc = rc or MAX_VERSION
+        self.post1 = post1 or MAX_VERSION
+        self.post2 = post2 or MAX_VERSION
+        self.dev = dev or MAX_VERSION
+        self._release_specificity = release_specificity
+
+    @property
+    def as_tuple(self) -> Tuple[Union[int, float], ...]:
+        return (
+            self.epoch,
+            self.major,
+            self.minor,
+            self.patch,
+            self.alpha,
+            self.beta,
+            self.rc,
+            self.post1,
+            self.post2,
+            self.dev,
+        )
+
+    @classmethod
+    def from_string(cls, raw: str, /) -> "Constraint":
+        if not (match := CONSTRAINT_PATTERN.match(raw)):
+            raise ValueError("invalid version string")
+
+        version = match.groupdict()
+        raw_release = version["release"]
+
+        if "*" in raw_release:
+            version["comparitor"] = "~="
+            raw_release = raw_release.replace("*", "0")
+
+        release = [int(v) for v in raw_release.split(".")]
+        n_parts = len(release)
+        release += [0] * (3 - n_parts)
+
+        def maybe_int(value: Optional[str]) -> Optional[int]:
+            return int(value) if value else None
+
+        kwargs = (
+            {version["pre_l"]: maybe_int(version["pre_n"])} if version["pre"] else {}
+        )
+
+        return cls(
+            version["comparitor"],
+            epoch=maybe_int(version["epoch"]),
+            major=release[0],
+            minor=release[1],
+            patch=release[2],
+            post1=maybe_int(version["post_n1"]),
+            post2=maybe_int(version["post_n2"]),
+            dev=maybe_int(version["dev_n"]),
+            release_specificity=n_parts,
+            **kwargs,
+        )
+
+    def _tilde_callback(
+        self,
+        constraint: Tuple[Union[int, float], ...],
+        requirement: Tuple[Union[int, float], ...],
+    ) -> bool:
+        rs = self._release_specificity
+        return constraint[:rs] == requirement[:rs] and requirement[rs] >= constraint[rs]
+
+    def likes_version(self, version: str, /) -> bool:
+        other = Constraint.from_string(f"=={version}")
+        COMPARITOR_MAPPING = {
+            "==": self.as_tuple.__eq__,
+            "!=": self.as_tuple.__ne__,
+            ">": self.as_tuple.__lt__,
+            ">=": self.as_tuple.__le__,
+            "<": self.as_tuple.__gt__,
+            "<=": self.as_tuple.__ge__,
+            "~=": partial(self._tilde_callback, self.as_tuple),
+        }
+        return COMPARITOR_MAPPING[self.comparitor](other.as_tuple)
+
+
+class Dependency:
+    def __init__(
+        self,
+        name: Optional[str] = None,
+        *,
+        constraints: Optional[List[str]] = None,
+        env_markers: Optional[List[str]] = None,
+        extras: Optional[List[str]] = None,
+        location: Optional[Union[str, Path]] = None,
+        editable: bool = False,
+    ) -> None:
```

Comment:
This should probably go to a new file, as all parsers will resolve to a list of these dependencies.